cirq-core/cirq/transformers/heuristic_decompositions/two_qubit_gate_tabulation.py

```diff
@@ -0,0 +1,496 @@
+    #    KAK vector.
+    missed_points = []
+    base_gate_dag = base_gate.conj().T
+    for ind in missing_vec_inds:
+        missing_vec = mesh_points[ind]
+        # Unitary A we wish to solve for
+        missing_unitary = kak_vector_to_unitary(missing_vec)
+
+        # Products of the from base_gate^\dagger k A
+        products = np.einsum('ab,...bc,cd', base_gate_dag, u_locals, missing_unitary)
+        # KAK vectors for these products
+        kaks = cirq.kak_vector(products, check_preconditions=False)
+        kaks = kaks[..., np.newaxis, :]
+
+        # Check if any of the product KAK vectors are close to a previously
+        # tabulated KAK vector
+        dists2 = np.sum((kaks - kak_vecs_single) ** 2, axis=-1)
+        min_dist_inds = np.unravel_index(dists2.argmin(), dists2.shape)
+        min_dist = np.sqrt(dists2[min_dist_inds])
+        if min_dist < tabulation_cutoff:
+            # If so, compute the single qubit unitary k_L such that
+            # base_gate^\dagger k A = kL base_gate k0 base_gate kR
+            # where k0 is the old (previously tabulated) single qubit unitary
+            # and k is one of the single qubit unitaries used above.
+            # Indices below are for k, k0 respectively
+            new_ind, old_ind = min_dist_inds
+
+            # Special case where the RHS is just base_gate (no single qubit
+            # gates yet applied). I.e. base_gate^\dagger k A ~  base_gate
+            # which implies  base_gate^\dagger k A = k_L base_gate k_R
+            new_product = products[new_ind]
+            if old_ind == 0:
+                assert not sq_cycles_single[old_ind]
+                base_product = base_gate
+                _, kL, actual = _outer_locals_for_unitary(new_product, base_product)
+                # Add to the enumeration
+                sq_cycles.append((kL,))
+            else:  # typical case mentioned above
+                assert len(sq_cycles_single[old_ind]) == 1
+                old_sq_cycle = sq_cycles_single[old_ind][0]
+                old_k = np.kron(*old_sq_cycle)
+                base_product = base_gate @ old_k @ base_gate
+                _, kL, actual = _outer_locals_for_unitary(new_product, base_product)
+                # Add to the enumeration
+                sq_cycles.append((old_sq_cycle, kL))
+
+            kak_vecs.append(cirq.kak_vector(base_gate @ actual, check_preconditions=False))
+        elif not allow_missed_points:
+            raise ValueError(f'Failed to tabulate a KAK vector near {missing_vec}')
+        else:
+            missed_points.append(missing_vec)
+
+    kak_vecs = np.array(kak_vecs)
+    summary += (
+        f'\nFraction of Weyl chamber reached with 2 gates and 3 gates '
+        f'(after patchup)'
+        f': {(len(kak_vecs) - 1) / num_mesh_points :.3f}'
+    )
+
+    return TwoQubitGateTabulation(
+        base_gate, kak_vecs, sq_cycles, max_infidelity, summary, tuple(missed_points)
+    )
+
+
+# pylint: enable=missing-raises-doc
```

Comment:
Now that we have the copyright in place, this line is not needed. Please delete.

Reply:
Done.